src/olympia/git/models.py

```diff
@@ -0,0 +1,26 @@
+from django.db import models
+
+from olympia.addons.models import Addon
+from olympia.amo.models import ModelBase
+
+
+class AddonGitExtraction(ModelBase):
+    """
+    This is an add-on related model that stores information related to the git
+    extraction (for code-manager).
+    """
+
+    addon = models.OneToOneField(
+        Addon, on_delete=models.CASCADE, primary_key=True
+    )
+    in_progress = models.BooleanField(default=False)
+
+
+class GitExtractionEntry(ModelBase):
```

Comment:
Actually, I'm not sure why we have two models that are both fks to `Addon` and both for git extraction.  How would adding the `in_progress` field to this model instead change things?

Reply:
The other model is there to avoid adding a new field to the `Addon` model (1-1) while this model (`GitExtractionEntry`) can have the same add-on multiple times and we create and delete entries in this table (that acts as a sort of queue).

Reply:
There would be multiple instances of the same add-on in the table at the same time?  Why would you need that?  (surely the first time would extract all the versions)

Reply:
There is no guarantee that all the versions will be extracted by the same CRON task execution. Because the CRON is scheduled at a fixed interval and submissions are random, we could likely have a scenario where a new version is submitted after the CRON retrieved the versions to extract.

Reply:
I've been thinking the design over and I still don't think two models are required.  If `in_progress` was added to this model then the only change (that I can see) would be change to `Addon.git_extraction_is_in_progress` to

```
    def git_extraction_is_in_progress(self):
        return GitExtractionEntry.objects.filter(addon=self, in_progress=True).exists()
```

I understand this came off the back of the pr discussions with @diox though and I've not trawled through all of that (I made it half way through but there were too many collapsed threads!)

Reply:
Here is a scenario:

1. the `git_extraction` CRON task starts 10 tasks to extract the different versions from v1 to v10, those versions are big so it takes more than 5 minutes to extract them
2. user submits new version: v11
3. v1 to v6 have been extracted so far
4. the `git_extraction` CRON task starts again

How do we prevent the second set of tasks to no run before the task for v10? To me, the answer is: locking the add-on so that the task is not created for v11. It will eventually be created once the add-on is unlocked.

So now, how do we remember that v11 needs extraction if we only use a single boolean flag?

This is why, I believe, we need two different models. The `AddonGitExtraction` model is really used to not add new columns to the `addons` table, otherwise the `git_extraction_in_progress` field would live in there. The purpose of this field is to prevent a git extraction of an add-on when one is already in progress, because we want to change our approach from: "let's git extract a version on the fly" to "let's git extract an add-on". In this case, extracting an add-on means extracting all the versions that haven't been extract yet, in the right order.

The `GitExtractionEntry` model is used to add add-on IDs in a table that will be consumed by the CRON task. If it's ok to create a rabbitmq to do this instead, I am fine with it but I don't think we do that.

With these two models, we can make sure that at step (4), the CRON task will skip the add-on because it is locked by `AddonGitExtraction.in_progress` but once this flag is set to `false`. In addition, at step (2) a new `GitExtractionEntry` will be created so that, once the add-on is unlocked, the CRONtask will create a task to extract all the not-yet-extracted versions (v11 here).

That being said, this is coming from my perspective and if we could make the scenario above work with a single model, then I'm fine with that because I don't really mind having one or more models. It's an implementation detail to me. The scenario above is important, though. WDYT?

Reply:
> How do we prevent the second set of tasks to no run before the task for v10? To me, the answer is: locking the add-on so that the task is not created for v11. It will eventually be created once the add-on is unlocked.

> So now, how do we remember that v11 needs extraction if we only use a single boolean flag?

`GitExtractionEntry.objects.filter(addon=self, in_progress=True).exists()` would be true (via `Addon.git_extraction_is_in_progress`).  You'd only need a separate model if you intended to persist the flag beyond the time the task was waiting to be processed and being processed, but as I'm understanding the code and issues/comments the flag is set back to `in_progress=False` which is the same as not existing.

Reply:
Yeah I agree with @eviljeff , I think one model is enough. To ensure consistency you'd need an unique constraint on (`addon`, `in_progress`), and default `in_progress` to `null` (`NullBooleanField`) and never set it to `False`, only `True` when an extraction is in progress.

Reply:
I am sorry, I don't understand, and it's certainly on me.

Do you mean not having a field on the `Addon` model (or the 1:1 with `AddonGitExtraction`) and relying on `GitExtractionEntry` to determine when an add-on is being extracted or not?

How would we retrieve the add-ons to process in the CRON task? And how would we mark them as "consumed"? (which means removing them from the queue because the git-extraction tasks have been created and submitted)

> To ensure consistency you'd need an unique constraint on (addon, in_progress)

Wouldn't that fail (the second time) if two versions of the same add-on are submitted while this add-on is being git-extracted?

> and default in_progress to null (NullBooleanField) and never set it to False, only True when an extraction is in progress.

Why never setting it back to `False`?

Reply:
If you set it to `null` instead of `False` then you can have multiple entries with the same `addon`, the unique constraint would not get in the way. It would fail otherwise.

Reply:
> How would we retrieve the add-ons to process in the CRON task? And how would we mark them as "consumed"? (which means removing them from the queue because the git-extraction tasks have been created and submitted)

Take row by row, if `GitExtractionEntry.objects.filter(addon=self, in_progress=True).exists()` is `True` then skip it. When extraction is done remove the row you were processing (or just `GitExtractionEntry.objects.filter(addon=addon, in_progress=True).delete()` which should do the same thing since there should only be one extraction in progress per add-on at any given time)

Reply:
Ok, I think I see now. Sorry @eviljeff, if that was what you were saying from the beginning, then obviously I don't know how to read :/